appcues/src/main/java/com/appcues/qualification/DefaultQualificationService.kt

```diff
@@ -0,0 +1,241 @@
+import com.appcues.data.remote.appcues.request.EventRequest
+import com.appcues.data.session.PrefSessionLocalSource
+import com.appcues.logging.Logcues
+import com.appcues.util.doIfFailure
+import com.appcues.util.doIfSuccess
+import com.squareup.moshi.JsonDataException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import java.util.Date
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+
+internal class DefaultQualificationService(
+    private val appcuesRemoteSource: AppcuesRemoteSource,
+    private val appcuesLocalSource: AppcuesLocalSource,
+    private val sessionLocalSource: PrefSessionLocalSource,
+    private val experienceMapper: ExperienceMapper,
+    private val config: AppcuesConfig,
+    private val logcues: Logcues,
+) : QualificationService {
+
+    private val processingActivity: HashSet<UUID> = hashSetOf()
+    private val mutex = Mutex()
+
+    override suspend fun qualify(intents: List<AnalyticsIntent>): QualificationResult? {
+        val activityRequest = intents.mergeToActivityRequest()
+
+        // FIXME trying to recall reason why old call site flushPendingActivity could accept null as date.
+        SdkMetrics.tracked(activityRequest.requestId, Date())
+
+        return trackActivity(activityRequest).also {
+            // We can check here if there are experiences or else we just cancel the SdkMetrics track
+            if (it == null || it.experiences.isEmpty())
+                SdkMetrics.remove(activityRequest.requestId)
+        }
+    }
+
+    private suspend fun List<AnalyticsIntent>.mergeToActivityRequest(): ActivityRequest {
+        val events = filterIsInstance<Event>().map {
+            EventRequest(
+                name = it.name,
+                // FIXME attr and context wont need to be mutable anymore
+                attributes = it.attributes?.toMutableMap() ?: hashMapOf(),
+                context = it.context?.toMutableMap() ?: hashMapOf()
+            )
+        }
+
+        val profileUpdate = mutableMapOf<String, Any>().apply {
+            // Anonymous, UpdateProfile are also Identify at this point
+            filterIsInstance<Identify>().forEach {
+                it.properties?.run { putAll(this) }
+            }
+        }
+
+        val groupUpdate = mutableMapOf<String, Any>().apply {
+            filterIsInstance<UpdateGroup>().forEach {
+                it.properties?.run { putAll(this) }
+            }
+        }
+
+        return ActivityRequest(
+            accountId = config.accountId,
+            // at this point it should never be null
+            userId = sessionLocalSource.getUserId() ?: String(),
+            profileUpdate = profileUpdate,
+            groupId = sessionLocalSource.getGroupId(),
+            groupUpdate = groupUpdate,
+            events = events,
+            userSignature = sessionLocalSource.getUserSignature(),
```

Comment:
the usage of `sessionLocalSource` here makes me possibly uneasy - what if the session data changed from the moment when this event was tracked to the moment here where we are sending it - is that possible? Would it be possible to capture that session data into the AnalyticsIntent earlier? at the moment of initial processing? that way each analytics item will be an atomic set of data that we know matches truly to the time it was captured.

Reply:
I think if we make sure that when we call "process" we know it will wait until its done to call the next then we make sure there is no way information will change, right?

Reply:
yeah probably so - I'd feel most comfortable if the values were immediately stuck into the AnalyticsIntent, so that we are guaranteed that nothing downstream could affect the user/signature/group information.

Reply:
thats one of the pieces that surely needs to change 😭 , because we queue Intents on the main thread which makes sense, but we cant decorate it while doing so.

Reply:
its probably better now!